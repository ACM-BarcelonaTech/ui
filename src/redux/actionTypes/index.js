const CELLS = 'CELLS';
const LOAD_CELLS = `${CELLS}.LOAD`;

const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const DIFF_EXPR = 'DIFF_EXPR';
const LOAD_DIFF_EXPR = `${DIFF_EXPR}.LOAD`;
const UPDATE_DIFF_EXPR = `${DIFF_EXPR}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const UPDATE_CELL_INFO = 'UPDATE_CELL_INFO';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_CELLS,
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
  UPDATE_CELL_INFO, SET_FOCUSED_GENE,
};
