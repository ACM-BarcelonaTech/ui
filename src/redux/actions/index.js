/* eslint-disable no-param-reassign */
import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, PUSH_CELL_SETS, CREATE_CLUSTER, CELL_SETS_COLOR,
  UPDATE_GENE_LIST, LOAD_GENE_LIST, SELECTED_GENES, UPDATE_GENE_EXPRESSION,
  LOAD_CELLS, BUILD_HEATMAP_SPEC, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
} from './actionType';
import connectionPromise from '../../utils/socketConnection';

const loadCellSets = (experimentId) => (dispatch, getState) => {
  if (getState().cellSets.data) {
    return null;
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      experimentId,
      data: json.cellSets,
    }),
  ).catch((e) => console.error('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (experimentId, newState) => (dispatch, getState) => {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  dispatch({
    type: UPDATE_CELL_SETS,
    experimentId,
    data: newState,
  });

  return dispatch(pushCellSets(experimentId));
};

const pushCellSets = (experimentId) => (dispatch, getState) => {
  // If no loaded data exists for our cell sets and some event would
  // trigger a push, we do not want to overwrite our cell sets with
  // empty data.
  if (!getState()?.cellSets?.data) {
    return null;
  }

  return fetch(
    `${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(
        getState().cellSets.data,
        (k, v) => ((k === 'title') ? undefined : v),
      ),
    },
  ).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: PUSH_CELL_SETS,
      experimentId,
      data: json,
    }),
  ).catch((e) => console.error('Error when trying to push cell sets to API: ', e));
};

const createCluster = (experimentId, cellSetInfo, clusterName, color) => (dispatch) => {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  dispatch({
    type: CREATE_CLUSTER,
    experimentId,
    data: newCluster,
  });

  return dispatch(pushCellSets(experimentId));
};

const cellSetsColor = (colorData) => (dispatch, getState) => {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return null;
};

const loadCells = (experimentId, embeddingType) => (dispatch, getState) => {
  if (getState().cells.data) {
    return null;
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const body = {
      name: 'GetEmbedding',
      type: embeddingType,
    };

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      return dispatch({
        type: LOAD_CELLS,
        experimentId,
        data: embedding,
      });
    });
  });
};

const updateGeneList = (experimentId, tableState) => (dispatch, getState) => {
  if (getState().geneList?.tableState === tableState) {
    return null;
  }
  dispatch({
    type: LOAD_GENE_LIST,
  });
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();
    const { geneList } = getState();

    const orderBy = tableState?.sorter.field
      || geneList.tableState?.sorter.field
      || 'dispersions';

    const orderDirection = tableState?.sorter.order
      || geneList.tableState?.sorter.order
      || 'ascend';

    const currentPage = tableState?.pagination.current
      || geneList.tableState?.pagination.current
      || 1;

    const currentPageSize = tableState?.pagination.pageSize
      || geneList.tableState?.pagination.pageSize
      || 1;

    const body = {
      name: 'ListGenes',
      selectFields: ['gene_names', 'dispersions'],
      orderBy,
      orderDirection: (orderDirection === 'ascend') ? 'ASC' : 'DESC',
      offset: ((currentPage - 1) * currentPageSize),
      limit: currentPageSize,
    };

    if (tableState.geneNamesFilter) {
      body.geneNamesFilter = tableState.geneNamesFilter;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);
    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const data = JSON.parse(res.results[0].body);

      const { total, rows } = data;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      if (tableState && tableState.pagination) {
        tableState.pagination.total = total;
      }
      return dispatch({
        type: UPDATE_GENE_LIST,
        experimentId,
        data: {
          rows,
          tableState: tableState || geneList.tableState,
        },
      });
    });
  });
};

const loadDiffExpr = (
  experimentId, comparisonType, firstSelectedCluster, secondSelectedCluster,
) => (dispatch) => {
  dispatch({
    type: LOAD_DIFF_EXPR,
  });
  return connectionPromise().then((io) => {
    const ComparisonTypes = {
      One: 'Versus Rest',
      Two: 'Across Sets',
    };

    const requestUuid = uuidv4();

    const body = {
      name: 'DifferentialExpression',
      maxNum: 100,
      cellSet: firstSelectedCluster.key,
    };
    if (comparisonType === ComparisonTypes.One) {
      body.compareWith = 'rest';
    } else {
      body.compareWith = secondSelectedCluster.key;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      let data = {};

      try {
        data = JSON.parse(res.results[0].body);
      } catch (error) {
        console.error(error);
        data = { rows: [] };
      }

      const { rows } = data;
      const total = rows.length;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      return dispatch({
        type: UPDATE_DIFF_EXPR,
        experimentId,
        data: {
          allData: rows,
          total,
        },
      });
    });
  });
};

const updateSelectedGenes = (genes, selected) => (dispatch, getState) => {
  const { selectedGenes, geneExperessionData } = getState();
  selectedGenes.geneList = selectedGenes.geneList || {};
  let newGenesAdded = false;
  genes.forEach((gene) => {
    if (selected) {
      if (!selectedGenes.geneList[gene]) {
        // This will present a dict of newly selected genes from gene table
        // { <Gene Name>: <Is the gene experession resolved?> }
        selectedGenes.geneList[gene] = false;
        newGenesAdded = true;
      }
    } else {
      delete selectedGenes.geneList[gene];
      const foundGene = geneExperessionData.data.findIndex((g) => g.geneName === gene);
      geneExperessionData.data.splice(foundGene, 1);
    }
  });
  dispatch({
    type: SELECTED_GENES,
    data: { newGenesAdded },
  });
  dispatch({
    type: UPDATE_HEATMAP_SPEC,
    data: {
      genes: geneExperessionData.data,
      rendering: true,
    },
  });
  setTimeout(() => {
    // Running this action with a delay to allow rerendering of heatmap
    dispatch({
      type: UPDATE_HEATMAP_SPEC,
      data: { rendering: false },
    });
  }, 50);
};

const loadGeneExpression = (experimentId) => (dispatch, getState) => {
  const { geneExperessionData } = getState();
  if (geneExperessionData?.isLoading) {
    return null;
  }
  const { newGenesAdded, geneList } = getState().selectedGenes;
  const newGeneBatch = [];
  if (newGenesAdded) {
    // eslint-disable-next-line no-restricted-syntax
    for (const [key, value] of Object.entries(geneList)) {
      if (!value) {
        newGeneBatch.push(key);
        geneList[key] = true;
      }
    }
  }
  dispatch({
    type: SELECTED_GENES,
    data: {
      newGenesAdded: false,
    },
  });
  if (newGeneBatch.length) {
    dispatch({
      type: UPDATE_GENE_EXPRESSION,
      data: {
        isLoading: true,
      },
    });
    return connectionPromise().then((io) => {
      const requestUuid = uuidv4();

      const body = {
        name: 'GeneExpression',
        cellSets: 'all',
        genes: newGeneBatch,
      };

      const request = {
        uuid: requestUuid,
        socketId: io.id,
        experimentId,
        timeout: '2021-01-01T00:00:00Z',
        body,
      };

      io.emit('WorkRequest', request);

      io.on(`WorkResponse-${requestUuid}`, (res) => {
        const heatMapData = JSON.parse(res.results[0].body);
        const { data } = getState().geneExperessionData;
        if (data) {
          Array.prototype.push.apply(heatMapData.data, data);
        }
        dispatch({
          type: UPDATE_GENE_EXPRESSION,
          data: {
            heatMapData,
            isLoading: false,
          },
        });
        return dispatch({
          type: BUILD_HEATMAP_SPEC,
          data: {
            geneExperessionData: heatMapData,
          },
        });
      });
    });
  }
};

export {
  loadCellSets,
  updateCellSets,
  pushCellSets,
  createCluster,
  loadCells,
  cellSetsColor,
  updateGeneList,
  updateSelectedGenes,
  loadGeneExpression,
  loadDiffExpr,
};
