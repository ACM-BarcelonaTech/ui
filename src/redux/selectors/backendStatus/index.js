import getBackendStatus from './getBackendStatus';

// eslint-disable-next-line import/prefer-default-export
export { getBackendStatus };
