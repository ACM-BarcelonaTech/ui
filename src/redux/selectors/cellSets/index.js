import getCellSets from './getCellSets';
import getCellSetsHierarchy from './getCellSetsHierarchy';
import getCellSetsHierarchyByType from './getCellSetsHierarchyByType';
import getCellSetsHierarchyByKey from './getCellSetsHierarchyByKey';

// eslint-disable-next-line import/prefer-default-export
export {
  getCellSets,
  getCellSetsHierarchy,
  getCellSetsHierarchyByType,
  getCellSetsHierarchyByKey,
};
