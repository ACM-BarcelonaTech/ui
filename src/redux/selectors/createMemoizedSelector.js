import { createSelector } from 'reselect';
import memoize from 'lru-memoize';

import _ from 'lodash';

const createMemoizedSelector = (
  selector, inputSelectors = [(state) => state], maxCachedKeys = 1,
) => {
  const makerFunction = (...params) => createSelector(
    inputSelectors,
    selector(...params),
  );

  return memoize(maxCachedKeys, _.isEqual)(makerFunction);
};

export default createMemoizedSelector;
