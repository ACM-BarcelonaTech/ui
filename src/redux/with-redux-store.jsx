/* eslint-disable no-underscore-dangle */
import React, { Component } from 'react';
import initializeStore from './store';

const __NEXT_REDUX_STORE__ = '__NEXT_REDUX_STORE__';

const getOrCreateStore = (initialState) => {
  // Always make a new store if server, otherwise state is shared between requests
  if (typeof window === 'undefined') {
    return initializeStore(initialState);
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[__NEXT_REDUX_STORE__]) {
    window[__NEXT_REDUX_STORE__] = initializeStore(initialState);
  }
  return window[__NEXT_REDUX_STORE__];
};

const withReduxStore = (App) => class AppWithRedux extends Component {
  static async getInitialProps(appContext) {
    // Get or Create the store with `undefined` as initialState
    // This allows you to set a custom default initialState
    const store = getOrCreateStore();

    // Provide the store to getInitialProps of pages
    // eslint-disable-next-line no-param-reassign
    appContext.ctx.store = store;

    return {
      ...(App.getInitialProps ? await App.getInitialProps(appContext) : {}),
      initialReduxState: store.getState(),
    };
  }

  render() {
    // eslint-disable-next-line react/prop-types
    const { initialReduxState } = this.props;
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <App {...this.props} store={getOrCreateStore(initialReduxState)} />;
  }
};

export default withReduxStore;
