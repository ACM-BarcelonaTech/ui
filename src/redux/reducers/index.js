/* eslint-disable no-param-reassign */
import { combineReducers } from 'redux';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, CREATE_CLUSTER, LOAD_CELLS,
  CELL_SETS_COLOR, UPDATE_GENE_LIST, LOAD_GENE_LIST,
  LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
} from '../actions/actionType';


const cellSetsReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_CELL_SETS:
      state.data = action.data;
      return state;
    case UPDATE_CELL_SETS:
      state.data = action.data;
      return state;
    case CREATE_CLUSTER:
      // for now, if cell set tool is not opened yet, we do nothing on create cell set action
      // in the future, we will need to handle that case too.
      if (!state.data) {
        // Find scratchpad at top level and add the new cluster.
        // The assignment is necessary because otherwise `useSelector`
        // won't recognize the changed state and the cell set tool won't rerender.

        state.data = state.data.map((topCategory) => {
          if (topCategory.key === 'scratchpad') {
            if (!topCategory.children) {
              topCategory.children = [];
            }

            topCategory.children.push(action.data);
          }

          return topCategory;
        });
      }
      return state;
    default:
      return state;
  }
};

const cellSetsColorReducer = (state = {}, action) => {
  switch (action.type) {
    case CELL_SETS_COLOR:
      state.data = action.data;
      return state;
    default:
      return state;
  }
};

const cellsReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_CELLS:
      state.data = action.data;
      return state;
    default:
      return state;
  }
};

const geneListReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_GENE_LIST:
      return {
        ...state,
        loading: true,
      };
    case UPDATE_GENE_LIST:
      return {
        ...state,
        ...action.data,
        loading: false,
      };
    default:
      return state;
  }
};

const diffExprReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_DIFF_EXPR:
      return {
        ...state,
        loading: true,
      };
    case UPDATE_DIFF_EXPR:
      return {
        ...state,
        ...action.data,
        loading: false,
      };
    default:
      return state;
  }
};

export default combineReducers({
  cellSets: cellSetsReducer,
  cellSetsColor: cellSetsColorReducer,
  cells: cellsReducer,
  geneList: geneListReducer,
  diffExpr: diffExprReducer,
});
