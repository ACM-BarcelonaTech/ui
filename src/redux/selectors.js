/* eslint-disable import/prefer-default-export */

// Accumulates all the different selectors

import * as backendSelectors from './selectors/backendStatus';
import * as cellSetsSelectors from './selectors/cellSets';

const getBackendStatus = (...params) => (state) => (
  backendSelectors.getBackendStatus(...params)(state.backendStatus));

const getCellSets = (...params) => (state) => {
  console.log('inside selector');
  return cellSetsSelectors.getCellSets(...params)(state.cellSets);
};

export {
  getBackendStatus,
  getCellSets,
};
