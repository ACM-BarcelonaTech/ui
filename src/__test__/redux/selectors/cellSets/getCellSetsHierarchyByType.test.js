import { getCellSetsHierarchyByType } from 'redux/selectors';
import mockCellSets from 'utils/tests/mockStores/cellSets';

describe('Get cell sets hierarchy selector by type test', () => {
  it('returns correctly by one key', () => {
    const returnedHierarchy = getCellSetsHierarchyByType('firstType')(mockCellSets());
    expect(returnedHierarchy).toMatchSnapshot();
  });
});
