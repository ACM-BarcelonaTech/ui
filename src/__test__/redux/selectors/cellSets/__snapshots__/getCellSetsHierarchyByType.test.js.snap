// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get cell sets hierarchy selector test Returns all if no type provided 1`] = `Array []`;

exports[`Get cell sets hierarchy selector test returns only selected cellset types 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "key": "test-1",
      },
      Object {
        "key": "test-2",
      },
    ],
    "key": "test",
    "name": "Test",
    "type": "firstType",
  },
]
`;
